kotlin/reflect/src/main/java/com/squareup/moshi/kotlin/reflect/KotlinJsonAdapter.kt

```diff
@@ -282,6 +379,277 @@ class KotlinJsonAdapterFactory : JsonAdapter.Factory {
 
     val nonTransientBindings = bindings.filterNotNull()
     val options = JsonReader.Options.of(*nonTransientBindings.map { it.name }.toTypedArray())
-    return KotlinJsonAdapter(constructor, bindings, nonTransientBindings, options).nullSafe()
+    return KotlinJsonAdapter(constructorData, bindings, nonTransientBindings, options).nullSafe()
+  }
+
+  private infix fun KmType?.isEqualTo(other: KmType?): Boolean {
+    return when {
+      this === other -> true
+      this != null && other != null -> {
+        abbreviatedType == other.abbreviatedType
```

Comment:
Note that this will imply that types which are represented with the same runtime type are different, if they have a different abbreviation. Not sure if you'd like this, but if you would, then never mind:
```
typealias ListOfString = List<String>
typealias MyList<T> = List<T>
```
KmType of `ListOfString` isEqualTo KmType of `MyList<String>` -> `false`

Reply:
Hmm, that's a good point. Since we're just doing this for checking equality of the parameter and property. I guess since it's a legal assignment we can drop the alias check. Done in 36f945b